toolchain/lex/lex.cpp

```diff
@@ -1297,15 +1231,158 @@ auto Lexer::LexFileEnd(llvm::StringRef source_text, ssize_t position) -> void {
   // The end-of-file token is always considered to be whitespace.
   NoteWhitespace();
 
-  // Close any open groups. We do this after marking whitespace, it will
-  // preserve that.
-  if (!open_groups_.empty()) {
-    CloseInvalidOpenGroups(TokenKind::Error, position);
-  }
-
   buffer_.AddToken({.kind = TokenKind::FileEnd,
                     .token_line = current_line(),
                     .column = ComputeColumn(position)});
+
+  // If we had any mismatched brackets, issue diagnostics and fix them.
+  if (has_mismatched_brackets_ || !open_groups_.empty()) {
+    DiagnoseAndFixMismatchedBrackets();
+  }
+}
+
+// If brackets didn't pair or nest properly, find a set of places to insert
+// brackets to fix the nesting, issue suitable diagnostics, and update the
+// token list to describe the fixes.
+auto Lexer::DiagnoseAndFixMismatchedBrackets() -> void {
+  llvm::SmallVector<std::pair<int, TokenizedBuffer::TokenInfo>> new_tokens;
```

Comment:
A comment on this to explain how it is used would seem helpful.

Reply:
Done.